community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/spi/ProcedureSignature.scala

```diff
@@ -19,11 +19,28 @@
  */
 package org.neo4j.cypher.internal.compiler.v3_0.spi
 
+import org.neo4j.cypher.internal.compiler.v3_0.executionplan.{READ_WRITE, READ_ONLY, InternalQueryType}
 import org.neo4j.cypher.internal.frontend.v3_0.symbols.CypherType
 
+sealed trait ProcedureMode {
+  val queryType : InternalQueryType
+  def call(ctx:QueryContext, signature:ProcedureSignature, args:Seq[Any]) : Iterator[Array[AnyRef]]
+}
+case object PROC_READ_ONLY extends ProcedureMode {
```

Comment:
Scala conventions for `object`s is camelCase; `procReadOnly`, but this could be an exception if we want to design procedures as a visibly different entity.